company-flow.el

```diff
@@ -0,0 +1,91 @@
+;;; company-flow.el --- Flow backend for company-mode  -*- lexical-binding: t -*-
+(require 'cl-lib)
+
+(defun company-flow-handle-signal (process _event)
+  (when (memq (process-status process) '(signal exit))
+    (let ((callback (process-get process 'company-flow-callback))
+          (prefix (process-get process 'company-flow-prefix)))
+      ;; (message "%S" (company-flow-parse-output (company-flow-get-output process)))
+      (funcall callback (->> process
+                             company-flow-get-output
+                             company-flow-parse-output
+                             (all-completions prefix))))))
+
+(defun company-flow-make-candidate (line)
+  (let* ((words (split-string line " "))
+         (text (car words))
+         (meta (mapconcat 'identity (cdr words) " ")))
+    (propertize text 'meta meta)))
+
+(defun company-flow-parse-output (output)
+  (mapcar 'company-flow-make-candidate
+          (split-string output "\n")))
+
+(defun company-flow-get-output (process)
+  "Get the complete output of PROCESS."
+  (with-demoted-errors "Error while retrieving process output: %S"
+    (let ((pending-output (process-get process 'company-flow-pending-output)))
+      (apply #'concat (nreverse pending-output)))))
+
+(defun company-flow-receive-checker-output (process output)
+  "Receive a syntax checking PROCESS OUTPUT."
+  (push output (process-get process 'company-flow-pending-output)))
```

Comment:
I borrowed this from flycheck. I don't know what guarantees company has in terms of how many in flight async requests to a given backend it could have at once. If it's only one then yeah, buffer local or global variables would probably be fine. 

I also like that it keeps the state only in the place that needs it, but I assume that I just need to get more used to buffer local/global variables in emacs